File: src/routes/tags.ts
import { Router, Request, Response } from "express";

import { checkLibrary } from "../lib/library";
import { asyncHandler } from "../middleware/asyncHandler";
import { ApiError } from "../types/api";
import { checkAuth } from "../lib/auth";
import { createTag, deleteTag, getTags } from "../lib/tags";

export const tagsRouter = Router();

tagsRouter.get(
  "/tag/:libraryId/:entityId",
  checkAuth,
  asyncHandler(async (req: Request, res: Response) => {
    const { libraryId, entityId } = req.params;

    if (isNaN(Number(libraryId)) || isNaN(Number(entityId))) {
      throw new ApiError(400, "Invalid library ID or entity ID");
    }

    const library = await checkLibrary(libraryId);

    const tags = await getTags(
      library.type,
      Number(entityId),
      req.headers.user_id ? Number(req.headers.user_id) : 0
    );

    res.json({
      status: true,
      tags,
    });
  })
);

tagsRouter.post(
  "/tag/:libraryId/:entityId",
  checkAuth,
  asyncHandler(async (req: Request, res: Response) => {
    const { libraryId, entityId } = req.params;
    const { tag } = req.body;

    if (typeof tag !== "string" || tag.length > 32) {
      throw new ApiError(400, "Invalid tag");
    }

    if (isNaN(Number(libraryId)) || isNaN(Number(entityId))) {
      throw new ApiError(400, "Invalid library ID or entity ID");
    }

    const library = await checkLibrary(libraryId);

    await createTag(
      library.type,
      Number(entityId),
      req.headers.user_id ? Number(req.headers.user_id) : 0,
      tag
    );

    res.json({
      status: true,
    });
  })
);

tagsRouter.delete(
  "/tag/:libraryId/:entityId/:tag",
  checkAuth,
  asyncHandler(async (req: Request, res: Response) => {
    const { libraryId, entityId, tag } = req.params;

    if (typeof tag !== "string" || tag.length < 1 || tag.length > 32) {
      throw new ApiError(400, "Invalid tag");
    }

    if (isNaN(Number(libraryId)) || isNaN(Number(entityId))) {
      throw new ApiError(400, "Invalid library ID or entity ID");
    }

    const library = await checkLibrary(libraryId);

    await deleteTag(
      library.type,
      Number(entityId),
      req.headers.user_id ? Number(req.headers.user_id) : 0,
      tag
    );

    res.json({
      status: true,
    });
  })
);

export default tagsRouter;

File: internal/api/handlers/tags.go
package handlers

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"

	"github.com/devourer/server/internal/auth"
	"github.com/devourer/server/internal/db/queries"
)

// ListTags handles GET /tag/:libraryId/:entityId
func (h *Handlers) ListTags(c *gin.Context) {
	libraryID, err := strconv.ParseInt(c.Param("libraryId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid library ID"})
		return
	}
	entityID, err := strconv.ParseInt(c.Param("entityId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid entity ID"})
		return
	}

	lib, err := queries.GetLibraryByID(h.DB, libraryID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Library not found"})
		return
	}

	userID, _ := c.Get(auth.CtxUserID)
	uid, _ := userID.(int64)

	tags, err := queries.ListUserTags(h.DB, uid, entityID, lib.Type)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"status": false, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"status": true, "tags": tags})
}

// CreateTag handles POST /tag/:libraryId/:entityId
func (h *Handlers) CreateTag(c *gin.Context) {
	libraryID, err := strconv.ParseInt(c.Param("libraryId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid library ID"})
		return
	}
	entityID, err := strconv.ParseInt(c.Param("entityId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid entity ID"})
		return
	}

	var body struct {
		Tag string `json:"tag"`
	}
	if err := c.ShouldBindJSON(&body); err != nil || len(body.Tag) == 0 || len(body.Tag) > 32 {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid tag"})
		return
	}

	lib, err := queries.GetLibraryByID(h.DB, libraryID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Library not found"})
		return
	}

	userID, _ := c.Get(auth.CtxUserID)
	uid, _ := userID.(int64)

	if err := queries.CreateUserTag(h.DB, uid, entityID, lib.Type, body.Tag); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"status": false, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"status": true})
}

// DeleteTag handles DELETE /tag/:libraryId/:entityId/:tag
func (h *Handlers) DeleteTag(c *gin.Context) {
	libraryID, err := strconv.ParseInt(c.Param("libraryId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid library ID"})
		return
	}
	entityID, err := strconv.ParseInt(c.Param("entityId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid entity ID"})
		return
	}
	tag := c.Param("tag")
	if len(tag) == 0 || len(tag) > 32 {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid tag"})
		return
	}

	lib, err := queries.GetLibraryByID(h.DB, libraryID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Library not found"})
		return
	}

	userID, _ := c.Get(auth.CtxUserID)
	uid, _ := userID.(int64)

	if err := queries.DeleteUserTag(h.DB, uid, entityID, lib.Type, tag); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"status": false, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"status": true})
}
